PR to start support for Skydive integration with Ansible

justjais opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Skydive is a network analyzer tool, which helps in analyzing SDNs and network and with this PR Ansible API and lookup modules are added to query Skydive objects using skydive python client. Opened new PR closing the earlier skydive PR #50136
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
skydive
##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

justjais committed:
skydive modules

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
skydive modules

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
skydive modules

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
skydive modules

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
skydive modules

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
skydive modules

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
skydive modules

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

rcarrillocruz reviewed (commented):
The skydive module name seems odd to me.
You name it 'add_node', yet you provide a state present/absent.

Why not calling it 'skydive_node', which then manages the lifecycle of the skydive  'object' ?

justjais committed:
fix shippable

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
modifying file name

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
fix shippable

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
renamed

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
skydive lookup

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
change in str

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
change in str

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
fix shippable error

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
renamed file

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
fix shippable error

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
change in str

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
fix shippable error

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
fix shippable error

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais commented:
@safchain @pierrecregut please review the respective PR.

pierrecregut commented on lib/ansible/module_utils/network/skydive/skydive.py:
```diff
@@ -0,0 +1,171 @@
+    ''' Base class for implementing Skydive Rest API '''
+    provider_spec = {'provider': dict(type='dict', options=SKYDIVE_PROVIDER_SPEC)}
+
+    def __init__(self, host="localhost:8082", username="admin", password="password"):
```
No support for ssl_verify. Limited to http because you do not override scheme

justjais replied:
shall override scheme in another PR linked to node and edge module support.

pierrecregut commented on lib/ansible/module_utils/network/skydive/skydive.py:
```diff
@@ -0,0 +1,171 @@
+# USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+from ansible.module_utils.six import iteritems
```
Not very important but there are unused imports and flake8 complains about module order.

justjais replied:
resolved.

pierrecregut commented on lib/ansible/module_utils/network/skydive/skydive.py:
```diff
@@ -0,0 +1,171 @@
+SKYDIVE_LOOKUP_QUERY = 'G.V().Has'
+
+SKYDIVE_PROVIDER_SPEC = {
+    'host': dict(fallback=(env_fallback, ['SKYDIVE_HOST'])),
```
As it contains a port specification endpoint is probably a better name

justjais replied:
changed

pierrecregut commented on lib/ansible/module_utils/network/skydive/skydive.py:
```diff
@@ -0,0 +1,171 @@
+    def __init__(self, host="localhost:8082", username="admin", password="password"):
+        try:
+            self.restclient_object = RESTClient(host, username=username, password=password)
+            self.websocketclient_object = WSClient(host, username=username, password=password)
```
WSClient takes two mandatory arguments : the endpoint specification is the second argument. Anyway you never use it but you should if skydive_node as its name suggests adds nodes.

justjais replied:
removed the WSClient from this PR, will create a new PR with WSClient and support for node and edge feature

pierrecregut commented on lib/ansible/module_utils/network/skydive/skydive.py:
```diff
@@ -0,0 +1,171 @@
+        super(skydive_lookup, self).__init__(host, username, password)
+        self.query_str = ""
+
+    def lookup_query(self, filter_data):
```
Limited to nodes and limited to a single Has filter. I would not constrain the gremlin query that can be sent. If you want to
provide a simpler lookup for people who do not want to understand gremlin, provide it as an additional query. Example of things that are useful in practice : Find the interface node (type "interface") named "eth0' that is a child of a host (type "host") that is named "my_machine". When you glue nodes created by ansible and nodes found by skydive, you need at least this expressive power in your lookup queries.

justjais replied:
cool, got the use case and shall raise a new PR supporting the chained queries.

safchain replied:
I think that the filtering mechanism is a bit limited, maybe adding a chained queries will help but as Pierre suggested I would prefer to let the user uses a raw gremlin expression as they can be complex returning either nodes or edges, few examples :

```
G.V().Has('Type', 'veth', 'Name', 'eth0').In().Has('Name', 'vm1')
G.V().HasEither('Type', 'ovsbridge', 'Name', 'br-ex').Out('State', NE('UP'))
G.V().HasKey('MTU')
G.E().Has('RelationType', 'layer2').InV().Has('Name', 'en0')
```
A more complete Skydive gremlin documentation here : 
http://skydive.network/documentation/api-gremlin

Do you have an example of how such request will be handled with the chained queries ?

pierrecregut commented on lib/ansible/module_utils/network/skydive/skydive.py:
```diff
@@ -0,0 +1,171 @@
+        return nodes[0].__dict__
+
+
+class skydive_flow_topology(skydive_base):
```
Supporting captures is great, you may also want to support injecting traffic.

justjais replied:
Noted, shall add the support in new PR

pierrecregut commented on lib/ansible/module_utils/network/skydive/skydive.py:
```diff
@@ -0,0 +1,171 @@
+        obj_filter = dict([(k, self.module.params[k]) for k, v in iteritems(ib_spec) if v.get('ib_req')])
+
+        proposed_object = {}
+        for key, value in iteritems(ib_spec):
```
for key in iterkeys(ib_spec):

justjais replied:
done

pierrecregut commented on lib/ansible/module_utils/network/skydive/skydive.py:
```diff
@@ -0,0 +1,171 @@
+        return result
+
+
+class skydive_graph_engine(skydive_base):
```
I suspect it is more or less a placeholder. It does not do what it pretends to do. This is where you should use the WSClient

justjais replied:
I shall create a new PR with WSClient support which will ultimately support node and edge features

pierrecregut commented on lib/ansible/modules/network/skydive/skydive_node.py:
```diff
@@ -0,0 +1,107 @@
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
```
use six.add_metaclass for python3 compability tells flake8.

pierrecregut commented on lib/ansible/modules/network/skydive/skydive_node.py:
```diff
@@ -0,0 +1,107 @@
+    module = AnsibleModule(argument_spec=argument_spec,
+                           supports_check_mode=True)
+
+    skydive_obj = skydive_graph_engine()
```
You forget to pass module. Never tested ?

justjais replied:
removing this file from this PR, shall raise a new PR with node and edge module support asap

pierrecregut commented on lib/ansible/utils/module_docs_fragments/skydive.py:
```diff
@@ -0,0 +1,44 @@
+    suboptions:
+      host:
+        description:
+          - Specifies the host name or address for connecting to the remote
```
Name is not clear and description does not really help to understand the format hostname:port

justjais replied:
updated the description to better give the idea of hostname:port

pierrecregut commented:
I will add some more generic feedback on how to add nodes and edges to skydive and why some items are necessary.
* complex node lookup (I already explained why).
* WSClient host_id : there can only be one connection with a given host_id at a point in time. As by default ansible parallelizes tasks, you need to provide different host_ids.
* We need to give some control on the uuid in use. When you have a hierarchy of nodes with different levels, you often have items at the low level with the same name (eg: two interfaces "eth0" but on different hosts, or even on different cards of the same host, etc), if you use only your hostname (something like 'ansible') and the element name, you will have uuid collisions. 
* Creating a new websocket for each new node can be very slow if you have a lot of nodes. If you have an inventory manager, you probably have a way to get your complete switch topology with only one request, then instantiating it in Skydive may take half an hour on a moderate deployment. Using batch specification where you reuse the websocket can speed up node creation (in my case the full script had a 25x speed improvement)
* Having a lookup could replace the option in current skydive ansible module where you give gremlin queries as parameters to skydive_edge. But in this case, we probably need shortcuts to get directly the uuid and get one and only one uuid from a query without dozen lines of map and other weird filters.

pierrecregut commented on lib/ansible/plugins/lookup/skydive.py:
```diff
@@ -0,0 +1,77 @@
+#
```
Having a lookup module is a much better idea than the existing `skydive_query` module.

justjais replied:
+1

justjais committed:
fix review comments

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
fix review comments

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
deleting file to add to new PR

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
fix review comments

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais commented:
@pierrecregut Thanks for your review, and I have updated the respective PR with changes asked. Also, I have removed the support for WSClient and node module for the integration from this particular PR and shall open a new PR with added support for WSClient specific module i.e. node and edge, also shall add support for chained lookup queries as you have suggested in your review comment.

So, if you are ok and approve current PR changes I shall go ahead and merge this PR in ansible repo.

pierrecregut commented:
@justjais it would be better if you could wait for one of the Skydive core devs review. I have also only superficially looked at the capture code so far. 
I agree with almost all your changes but I am not completely convinced that you should postpone the support for insecure and scheme in the provider description for the Rest client. Even for creating captures, you will need it if the analyzer uses https (mandatory in production) and has self signed certificates (unfortunately often the case).

samccann commented on lib/ansible/modules/network/skydive/skydive_capture.py:
```diff
@@ -0,0 +1,136 @@
+  - This module manages flow capture on interfaces. The Gremlin
+    expression is continuously evaluated which means that it is
+    possible to define a capture on nodes that do not exist yet.
+  - It useful when you want to start a capture on all OpenvSwitch
```
```suggestion
  - It is useful when you want to start a capture on all OpenvSwitch
```

justjais replied:
done

samccann commented on lib/ansible/modules/network/skydive/skydive_capture.py:
```diff
@@ -0,0 +1,136 @@
+options:
+  name:
+    description:
+      - To define flow capture name
```
```suggestion
      - To define flow capture name.
```

justjais replied:
Done

samccann commented on lib/ansible/modules/network/skydive/skydive_capture.py:
```diff
@@ -0,0 +1,136 @@
+    required: true
+  capture_type:
+    description:
+      - To define flow capture type
```
```suggestion
      - To define flow capture type.
```

justjais replied:
Done

samccann commented on lib/ansible/modules/network/skydive/skydive_capture.py:
```diff
@@ -0,0 +1,136 @@
+        of this object.
+  extra_tcp_metric:
+    description:
+      - To define flow capture ExtraTCPMetric
```
```suggestion
      - To define flow capture ExtraTCPMetric.
```

justjais replied:
Done

samccann commented on lib/ansible/modules/network/skydive/skydive_capture.py:
```diff
@@ -0,0 +1,136 @@
+    default: false
+  ip_defrag:
+    description:
+      - To define flow capture IPDefrag
```
```suggestion
      - To define flow capture IPDefrag.
```

justjais replied:
Done

samccann commented on lib/ansible/modules/network/skydive/skydive_capture.py:
```diff
@@ -0,0 +1,136 @@
+    default: false
+  reassemble_tcp:
+    description:
+      - To define flow capture ReassembleTCP
```
```suggestion
      - To define flow capture ReassembleTCP.
```

justjais replied:
Done

samccann commented on lib/ansible/modules/network/skydive/skydive_capture.py:
```diff
@@ -0,0 +1,136 @@
+"""
+
+EXAMPLES = """
+- name: add a new new node with connected link
```
```suggestion
- name: add a new node with connected link
```

justjais replied:
Done

samccann commented on lib/ansible/plugins/lookup/skydive.py:
```diff
@@ -0,0 +1,77 @@
+short_description: Query Skydive objects
+description:
+  - Uses the Skydive python REST client to return the queried object from
+    Skydive network analyser.
```
```suggestion
    Skydive network analyzer.
```

justjais replied:
Done

samccann commented on lib/ansible/utils/module_docs_fragments/skydive.py:
```diff
@@ -0,0 +1,50 @@
+      endpoint:
+        description:
+          - Specifies the hostname/address along with the port as C(localhost:8082)for
+            connecting to the remote instance of SKYDIVE client over REST
```
```suggestion
            connecting to the remote instance of SKYDIVE client over the REST API.
```

justjais replied:
Done

samccann commented on lib/ansible/utils/module_docs_fragments/skydive.py:
```diff
@@ -0,0 +1,50 @@
+            the remote instance of SKYDIVE client.
+      verify_ssl:
+        description:
+          - Specifies the ssl parameter which decide the connection type shall be
```
```suggestion
          - Specifies the SSL parameter that decides if the connection type shall be
```

justjais replied:
Done

samccann commented on lib/ansible/utils/module_docs_fragments/skydive.py:
```diff
@@ -0,0 +1,50 @@
+      verify_ssl:
+        description:
+          - Specifies the ssl parameter which decide the connection type shall be
+            http/https.
```
```suggestion
            http or https.
```

justjais replied:
Done

samccann reviewed (commented):
some minor doc nits.

safchain commented on lib/ansible/module_utils/network/skydive/skydive.py:
```diff
@@ -0,0 +1,137 @@
+        return nodes[0].__dict__
+
+
+class skydive_flow_topology(skydive_restclient):
```
What do you think about skydive_flow_capture ? In skydive topology means interfaces, bridges, hosts, nodes here it is more about packet capture.

justjais committed:
shippable and doc fix

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
placing doc at correct path

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
shippable and doc fix

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
placing doc at correct path

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

safchain commented on lib/ansible/module_utils/network/skydive/skydive.py:
```diff
@@ -0,0 +1,137 @@
+                                                                obj_filter['ip_defrag'],
+                                                                obj_filter['reassemble_tcp'])
+            try:
+                self.restclient_object.capture_create(capture_str)
```
This won't work, the API call is incorrect, please see :
https://github.com/skydive-project/skydive/blob/master/contrib/python/api/skydive/rest/client.py#L122

safchain commented on lib/ansible/modules/network/skydive/skydive_capture.py:
```diff
@@ -0,0 +1,136 @@
+"""
+
+EXAMPLES = """
+- name: add a new new node with connected link
```
it should be more something like this:

```
start a new flow capture on the given node(s)
```

The gremlin query/expression is missing as parameter please see by upper comment.

safchain commented on lib/ansible/modules/network/skydive/skydive_capture.py:
```diff
@@ -0,0 +1,136 @@
+      username: admin
+      password: admin
+
+- name: remove node
```
should be something like:

```
Stop the flow capture
```

safchain commented on lib/ansible/plugins/lookup/skydive.py:
```diff
@@ -0,0 +1,77 @@
+
+- name: return skydive metdata if present based on TID
+  set_fact:
+    skydive: "{{ lookup('skydive_lookup', filter={'TID': '2b5e8263-89d3-5e01-506b-9120f49572b5'},
```
I tried to use/test it without success. I get the following error:

```
fatal: [localhost]: FAILED! => {"msg": "lookup plugin (skydive_lookup) not found"}
```

Is there a good (doc ?) way to deploy it properly so that we could test it and maybe create a CI job.

justjais committed:
updated with review and shippable fix

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
updated with review and shippable fix

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
updated with review and shippable fix

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
updated with review and shippable fix

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
fixing review comment

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
fixing review comment

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
fixing review comment

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
fixing review comment

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
fixing review comment

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
fix shippable errors

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
fix shippable error

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais commented:
@safchain @pierrecregut Apologies for the delay in updating the code with the review comments, I have implemented the suggested review comments. Kindly, review and let me know if it's ok to merge.

Also, as discussed earlier I shall raise different PR for supporting skydive edge and node modules.

pierrecregut commented on lib/ansible/module_utils/network/skydive/api.py:
```diff
@@ -0,0 +1,162 @@
+            if self.module.params[key] is not None:
+                proposed_object[key] = self.module.params[key]
+
+        capture_str = SKYDIVE_GREMLIN_QUERY + "('Name', '{0}', 'Type', '{1}', 'extra_tcp_metric', '{2}'\
```
There are two issues with this query.

As already pointed out by @safchain , you are mixing the arguments of the capture and the gremlin query which is one of those arguments. It characterizes the nodes where the capture is done. ip_defrag, reassemble_tcp, layer_key_mode are arguments of restclient_object.capture_create at line 143 and are not part of the gremlin query.

The second point is that the way you build the query is too restricted. Most of the time specifying a name and a type is not enough. For example in a data-center setting, with name 'eth0' and type 'Interface' you will get most of the primary interfaces of the machines of your data-center. You usually want the interface eth0 that is "owned" by a given host with a given name. You must let users run arbitrary query. You may supply simplified query, but those queries will be on UUID of nodes in Skydive. As its name implies it is a way to uniquely identify a node in Skydive and could be the outcome of a lot of commands (like creating a node as it is in the ansible module of the skydive team).

pierrecregut commented on lib/ansible/module_utils/network/skydive/api.py:
```diff
@@ -0,0 +1,162 @@
+                                                                                         obj_filter['layer_key_mode'])
+        # to check current object ref for idempotency
+        captured_list_objs = self.restclient_object.capture_list()
+        current_ref = False
```
Why not setting it to None and directly capturing the UUID. This would spare you the second pass for the delete case would not cost much in the creation case. You could/should also return this uuid as part of the result. As shown by the API, uuid are more trusted than names and users should be able to delete captures based on the uuid. In fact names are not mandatory.

pierrecregut commented on lib/ansible/module_utils/network/skydive/api.py:
```diff
@@ -0,0 +1,162 @@
+        return nodes[0].__dict__
+
+
+class skydive_flow_topology(skydive_restclient):
```
The name of the class does not reflect its intent.

justjais committed:
fix review comments

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
fixing shippable errors

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
fix review comments

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
review comments

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

justjais committed:
shippable fix

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

safchain commented on lib/ansible/plugins/lookup/skydive.py:
```diff
@@ -36,14 +36,14 @@
 - name: return skydive metdata if present based on TID
   set_fact:
-    skydive: "{{ lookup('skydive_lookup', filter={'query': \"G.V().Has('Name', 'sumit-VirtualBox')\"},
+    skydive: "{{ lookup('skydive', filter={'query': \"G.V().Has('Name')\"},
```
The query should be something like 

```
G.V().Has('TID', 'bd7f5334-115c-5943-487d-a77c486fa854')
```

according to the `name`

justjais replied:
updated the `name` field.

justjais committed:
review comment

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>
